File: Problema 2/C++/TestP2.cpp
// test_p2.cpp
#include "P2.cpp"
#include <iostream>
#include <vector>
#include <iomanip>

struct TestCase
{
    std::vector<int> l1;
    std::vector<int> l2;
    std::vector<int> expected;
    std::string description;
};

std::vector<int> listToVector(ListNode *head)
{
    std::vector<int> result;
    while (head != nullptr)
    {
        result.push_back(head->val);
        head = head->next;
    }
    return result;
}

// Función auxiliar para crear una lista enlazada a partir de un vector.
ListNode *createList(const std::vector<int> &nums)
{
    ListNode *dummyHead = new ListNode(0);
    ListNode *current = dummyHead;
    for (int num : nums)
    {
        current->next = new ListNode(num);
        current = current->next;
    }
    ListNode *result = dummyHead->next;
    delete dummyHead;
    return result;
}

// Función para imprimir una lista enlazada.
void printList(ListNode *head)
{
    while (head != nullptr)
    {
        std::cout << head->val;
        if (head->next != nullptr)
            std::cout << " -> ";
        head = head->next;
    }
    std::cout << std::endl;
}

void runTests()
{
    std::vector<TestCase> testCases = {
        {{2, 4, 3}, {5, 6, 4}, {7, 0, 8}, "342 + 465 = 807"},
        {{0}, {0}, {0}, "0 + 0 = 0"},
        {{9, 9, 9, 9, 9, 9, 9}, {9, 9, 9, 9}, {8, 9, 9, 9, 0, 0, 0, 1}, "9999999 + 9999 = 10009998"},
        {{1, 8}, {0}, {1, 8}, "81 + 0 = 81"},
        {{5}, {5}, {0, 1}, "5 + 5 = 10"}};

    Solution solution;
    int passed = 0;

    std::cout << "\nRunning " << testCases.size() << " tests:\n\n";

    for (size_t i = 0; i < testCases.size(); i++)
    {
        const auto &tc = testCases[i];
        ListNode *l1 = createList(tc.l1);
        ListNode *l2 = createList(tc.l2);
        ListNode *result = solution.addTwoNumbers(l1, l2);
        std::vector<int> actual = listToVector(result);

        bool success = actual == tc.expected;
        if (success)
        {
            std::cout << "\033[32m[PASS]\033[0m Test " << (i + 1) << ": " << tc.description << "\n";
            passed++;
        }
        else
        {
            std::cout << "\033[31m[FAIL]\033[0m Test " << (i + 1) << ": " << tc.description << "\n";
            std::cout << "       Expected: ";
            for (int n : tc.expected)
                std::cout << n << " ";
            std::cout << "\n       Got:      ";
            for (int n : actual)
                std::cout << n << " ";
            std::cout << "\n";
        }
    }

    double percentage = (passed * 100.0) / testCases.size();
    std::cout << "\nResults: " << passed << "/" << testCases.size()
              << " tests passed (" << std::fixed << std::setprecision(1)
              << percentage << "%)\n";
}

int main()
{
    runTests();
    return 0;
}

File: Problema 2/Ruby/TestP2.rb
# TestP2.rb
require_relative 'P2'
require 'test/unit'

# Métodos auxiliares para pruebas
def create_list(array)
  return nil if array.empty?
  auxiliar_node = ListNode.new
  actual_node = auxiliar_node
  array.each do |val|
    actual_node.next = ListNode.new(val)
    actual_node = actual_node.next
  end
  auxiliar_node.next
end

def list_to_array(head)
  result = []
  while head
    result << head.val
    head = head.next
  end
  result
end

class TestAddTwoNumbers < Test::Unit::TestCase
  def setup
    @test_cases = [
      {l1: [2,4,3], l2: [5,6,4], expected: [7,0,8], description: '342 + 465 = 807'},
      {l1: [0], l2: [0], expected: [0], description: '0 + 0 = 0'},
      {l1: [9,9,9,9,9,9,9], l2: [9,9,9,9], expected: [8,9,9,9,0,0,0,1], description: '9999999 + 9999 = 10009998'},
      {l1: [1,8], l2: [0], expected: [1,8], description: '81 + 0 = 81'},
      {l1: [5], l2: [5], expected: [0,1], description: '5 + 5 = 10'}
    ]
  end

  def test_add_two_numbers
    @test_cases.each do |tc|
      l1 = create_list(tc[:l1])
      l2 = create_list(tc[:l2])
      result = add_two_numbers(l1, l2)
      assert_equal(tc[:expected], list_to_array(result), "Failed for case: #{tc[:description]}")
    end
  end
end